skimage/filters/thresholding.py

```diff
@@ -726,6 +727,149 @@ def threshold_triangle(image, nbins=256):
+
+    Returns
+    -------
+    idx_thresh : (nclass) array
+        Array containing the threshold values for the desired classes.
+    max_sigma : float
+        Maximum sigma value achieved on the classes.
+
+    References
+    ----------
+    .. [1] Liao, P-S., Chen, T-S. and Chung, P-C., "A fast algorithm for
+           multilevel thresholding", Journal of Information Science and
+           Engineering 17 (5): 713-727, 2001. Available at:
+           <http://ftp.iis.sinica.edu.tw/JISE/2001/200109_01.pdf>
+    .. [2] Tosa, Y., "Multi-Otsu Threshold", a java plugin for ImageJ.
+           Available at:
+           <http://imagej.net/plugins/download/Multi_OtsuThreshold.java>
+
+    Examples
+    --------
+    >>> from skimage.color import label2rgb
+    >>> from skimage import data
+    >>> image = data.camera()
+    >>> thresh, _ = threshold_multiotsu(image)
+    >>> regions = np.digitize(image, bins=thresh)
+    >>> sep_regions = label2rgb(regions)
+    """
+    if image.shape[-1] in (3, 4):
+        raise TypeError("The input image seems to be RGB (shape: {0}. Please"
+                        "use a grayscale image.".format(image.shape))
+
+    if image.min() == image.max():
+        raise TypeError("The input image seems to have only one color: {0}."
+                        "Please use a grayscale image.".format(image.min()))
+
+    # check if nclass is between 2 and 5.
+    if nclass not in np.array((2, 3, 4, 5)):
+        raise ValueError("Please choose a number of classes between "
+                         "2 and 5.")
+
+    # receiving minimum and maximum values for the image type.
+    type_min, type_max = dtype_limits(image)
+
+    # calculating the histogram and the probability of each gray level.
+    hist, _ = np.histogram(image.ravel(), bins=nbins,
+                           range=(type_min, type_max))
+    prob = hist / image.size
+
+    max_sigma = 0
+    momP, momS, var_btwcls = [np.zeros((nbins, nbins)) for n in range(3)]
+
+    # building the lookup tables.
+    # step 1: calculating the diagonal.
+    for u in range(1, nbins):
+        momP[u, u] = prob[u]
+        momS[u, u] = u * prob[u]
+
+    # step 2: calculating the first row.
+    for u in range(1, nbins-1):
+        momP[1, u+1] = momP[1, u] + prob[u+1]
+        momS[1, u+1] = momS[1, u] + (u+1)*prob[u+1]
+
+    # step 3: calculating the other rows recursively.
+    for u in range(2, nbins):
+        for v in range(u+1, nbins):
+            momP[u, v] = momP[1, v] - momP[1, u-1]
+            momS[u, v] = momS[1, v] - momS[1, u-1]
+
+    # step 4: calculating the between class variance.
+    for u in range(1, nbins):
+        for v in range(u+1, nbins):
+            if (momP[u, v] != 0):
+                var_btwcls[u, v] = momS[u, v]**2 / momP[u, v]
+            else:
+                var_btwcls[u, v] = 0
+
+    # finding max threshold candidates, depending on nclass.
+    # number of thresholds is equal to number of classes - 1.
+    if nclass == 2:
```

Comment:
This would be pretty good, but I actually have no idea on how to walk within the `(idx[k]+1, bins-classes+k)` sets using `combinations()`. Could you clarify a little bit on how I could do that?